File: src/days/ts/day8.ts
let str = `30373
25512
65332
33549
35390`;

let data = str.split("\n");
let visibleTrees = 0;
let score = 0;

data.forEach((line,lineIndex) => {
    
    
    let numbers = line.trim().split("").map(str => Number.parseInt(str));
    if (lineIndex == 0) { 
         visibleTrees += numbers.length;
        return;
    };
    if(lineIndex + 1 == data.length) {
         visibleTrees += numbers.length;
        return;
    }
    numbers.forEach((number, index) => {
        if(index == 0) {
             visibleTrees++;
            return;
        };
        if(index + 1 == numbers.length) {
             visibleTrees++;
            return;
        }
      
        let numbersLeft = numbers.slice(0, index);
        let numbersRight = numbers.slice(index + 1, numbers.length);
        let numbersBottom = data.slice(lineIndex + 1, data.length).map(line => {
            let line2 = line.split("");
            return Number.parseInt(line2[index])
        });
        let numbersTop = data.slice(0, lineIndex).map(line => {
            let line2 = line.split("");
            return Number.parseInt(line2[index])
        });
        let isVisibleBottom = numbersBottom.filter(val => val >= number);
        let isVisibleTop = numbersTop.filter(val => val >= number);
        let isVisibleLeft = numbersLeft.filter(val => val >= number);
        let isVisibleRight = numbersRight.filter(val => val >= number);

        if (isVisibleBottom.length == 0 || isVisibleTop.length == 0 || isVisibleLeft.length == 0 || isVisibleRight.length == 0) {
            visibleTrees++;
        }
        const topViewingDistance = treeViewDistance(number, numbersTop.reverse());
        const BottomView = treeViewDistance(number, numbersBottom);
        const le = treeViewDistance(number, numbersLeft.reverse())
        const ri = treeViewDistance(number, numbersRight)
        const scn =  topViewingDistance * BottomView * le * ri
        // console.log(numbersRight)
        //  console.log(BottomView)
        // console.log(number)
        //  console.log()
        if(scn > score) score = scn
    })
})

function treeViewDistance(thisSpecificTree: number, treesInDirection: number[]) {
    let lastTree = 0;
    let viewingDistance = 0;

    for (let tree of treesInDirection) {

        if (tree >= lastTree || tree < thisSpecificTree) {

            viewingDistance++;
        }
        if(tree >= thisSpecificTree) break;
        lastTree = tree;
    }
    return viewingDistance;
}
console.log(visibleTrees)
console.log(score)

File: src/days/day8.rs
pub fn run(data: &String) {

    let lines = data.split("\n").collect::<Vec<_>>();

    let mut grid: Vec<Vec<u32>> = Vec::new();
    let mut visible_trees_part1 = 0;
    let mut highest_scenic_score_part2 = 0;
    
    // create a grid representation of the input using vectors and also convert them to integers. 
    for line in lines {
        let row = line.trim().chars().map(|char| char.to_digit(10).unwrap()).collect::<Vec<u32>>();
        grid.push(row);
    }

    for column_index in 0..grid[0].len() {
        let column = grid.iter().map(|row| row[column_index]).collect::<Vec<u32>>();
        
        
        // first column is always visible from the left side
        // last column is always visible from the right side.
        if column_index == 0 || column_index + 1 == grid[0].len() {
            visible_trees_part1 += column.len();
            continue;
        }

        for (row_index, tree) in column.iter().enumerate() {
            let row = &grid[row_index];
            
            // top most tree is always visible from the top
            // bottom most tree is always visible from the bottom
            if row_index == 0 || row_index + 1 == column.len() {
                visible_trees_part1 += 1;
                continue;
            }

            // using + 1 to skip the current tree.
            let bottom_trees = &column[row_index + 1..column.len()];
            let top_trees = &column[0..row_index];
            let left_trees = &row[0..column_index];
            let right_trees = &row[column_index + 1..row.len()];

            // part 1
            // check if the tree is visible from ANY one of the sides.
            if is_side_visible(tree, bottom_trees) || is_side_visible(tree, top_trees) || is_side_visible(tree, left_trees) || 
            is_side_visible(tree, right_trees) {
                visible_trees_part1 += 1;
            }

            // part 2
            // reverse top and left so its ordered FROM the tree
            let top_viewing_distance = tree_viewing_distance(tree, top_trees, true);
            let bottom_viewing_distance = tree_viewing_distance(tree, bottom_trees, false);
            let left_viewing_distance = tree_viewing_distance(tree, left_trees, true);
            let right_viewing_distance = tree_viewing_distance(tree, right_trees, false);

            let score = top_viewing_distance * bottom_viewing_distance * left_viewing_distance * right_viewing_distance;
            if score > highest_scenic_score_part2 {
                highest_scenic_score_part2 = score;
            }
        }
    }
    
    println!("part 1: {}", visible_trees_part1);
    println!("part 2: {}", highest_scenic_score_part2);
   
}

fn is_side_visible(tree: &u32, side_trees: &[u32]) -> bool {
    !side_trees.iter().any(|side_tree| side_tree >= tree)
}
fn tree_viewing_distance(tree: &u32, side_trees: &[u32], reverse: bool) -> u32 {
    let mut last_tree = 0;
    let mut viewing_distance = 0;
    
    let trees = if reverse {
        let mut copy_trees = side_trees.to_vec();
        copy_trees.reverse();
        copy_trees
    } else { side_trees.to_vec() };
    
    for side_tree in trees {
        if side_tree >= last_tree || side_tree < *tree {
            viewing_distance += 1;
        }
        if side_tree >= *tree { break; }
        last_tree = side_tree;
    }
    viewing_distance
}